src/nvim/lua/executor.c

```diff
@@ -397,6 +488,154 @@ static int nlua_wait(lua_State *lstate)
   return 2;
 }
 
+static void nlua_init_ref_state(nlua_ref_state_t *state)
+  FUNC_ATTR_NONNULL_ALL
+{
+  state->nil_ref = LUA_NOREF;
+  state->empty_dict_ref = LUA_NOREF;
+}
+
+static nlua_ref_state_t *nlua_get_ref_state(lua_State *lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  lua_getfield(lstate, LUA_REGISTRYINDEX, "nlua.ref_state");
+  nlua_ref_state_t *ref_state = lua_touserdata(lstate, -1);
+  lua_pop(lstate, 1);
+
+  return ref_state;
+}
+
+LuaRef nlua_get_nil_ref(lua_State *lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  nlua_ref_state_t *ref_state = nlua_get_ref_state(lstate);
+  return ref_state->nil_ref;
+}
+
+LuaRef nlua_get_empty_dict_ref(lua_State *lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  nlua_ref_state_t *ref_state = nlua_get_ref_state(lstate);
+  return ref_state->empty_dict_ref;
+}
+
+static int nlua_get_ref_count(lua_State *lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  nlua_ref_state_t *ref_state = nlua_get_ref_state(lstate);
+  return ref_state->ref_count;
+}
+
+int nlua_get_global_ref_count(void)
+{
+  lua_State *const lstate = global_lstate;
+  return nlua_get_ref_count(lstate);
+}
+
+static void nlua_common_vim_init(lua_State *lstate, bool is_thread)
+  FUNC_ATTR_NONNULL_ARG(1)
+{
+  nlua_ref_state_t *ref_state = xcalloc(1, sizeof(*ref_state));
+  nlua_init_ref_state(ref_state);
+  lua_pushlightuserdata(lstate, ref_state);
```

Comment:
a lightuserdata that "owns" an allocation is an anti-pattern. use a full userdata to store the `nlua_ref_state_t` directly instead.